src/ShallowWaterRosenbrock.jl

```diff
@@ -0,0 +1,243 @@
+  # energetically balanced second order rosenbrock shallow water solver
+  # reference: eqns (24) and (39) of
+  # https://github.com/BOM-Monash-Collaborations/articles/blob/main/energetically_balanced_time_integration/EnergeticallyBalancedTimeIntegration_SW.tex
+  #
+  # f          : coriolis force (field)
+  # g          : gravity (constant)
+  # h₁         : fluid depth at current time level
+  # u₁         : fluid velocity at current time level
+  # RTMM       : H(div) mass matrix, ∫β⋅βdΩ, ∀β∈ H(div,Ω)
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # τ          : potential vorticity upwinding parameter
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  dt₁ = dt
+  if leap_frog
+    dt₁ = 2.0*dt
+  end
+
+  # multifield terms
+  u₁, h₁ = y₁
+
+  # 1.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*h₁)
+  # 1.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,u₁⋅u₁,h₁,g)
+  # 1.3: the potential vorticity
+  compute_potential_vorticity!(q₁,H1h,H1hchol,dΩ,R,S,h₁,u₁,f,n)
+  # 1.4: assemble the velocity residual
+  bᵤ₁(v) = ∫(-1.0*(q₁ - τ*u₁⋅∇(q₁))*(v⋅⟂(F,n)))dΩ + ∫(DIV(v)*ϕ)dω
+  # 1.5: assemble the depth residual
+  bₕ₁(q) = ∫(-q*DIV(F))dω
+
+  bₕᵤ₁((v,q)) = bᵤ₁(v) + bₕ₁(q)
+  Gridap.FESpaces.assemble_vector!(bₕᵤ₁, get_free_dof_values(duh₁), Y)
+
+  # Solve for du₁, dh₁ over a MultiFieldFESpace
+  ldiv!(Blfchol, get_free_dof_values(duh₁))
+
+  # update
+  get_free_dof_values(y₂) .= get_free_dof_values(yₘ) .+ dt₁ .* get_free_dof_values(duh₁)
+
+  u₂, h₂ = y₂
+  # 2.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*(2.0*h₁ + h₂)/6.0+u₂*(h₁ + 2.0*h₂)/6.0)
+  # 2.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,(u₁⋅u₁ + u₁⋅u₂ + u₂⋅u₂)/3.0,0.5*(h₁ + h₂),g)
+  # 2.3: the potential vorticity
+  compute_potential_vorticity!(q₂,H1h,H1hchol,dΩ,R,S,h₂,u₂,f,n)
+  # 2.4: assemble the velocity residual
+  bᵤ₂(v) = ∫(-0.5*(q₁ - τ*u₁⋅∇(q₁) + q₂ - τ*u₂⋅∇(q₂))*(v⋅⟂(F,n)))dΩ + ∫(DIV(v)*ϕ)dω
+  # 2.5: assemble the depth residual
+  bₕ₂(q) = ∫(-q*DIV(F))dω
+
+  bₕᵤ₂((v,q)) = bᵤ₂(v) + bₕ₂(q)
+  Gridap.FESpaces.assemble_vector!(bₕᵤ₂, get_free_dof_values(duh₂), Y)
+
+  # subtract A*[du₁,dh₁] from [du₂,dh₂] vector
+  mul!(y_wrk, Amat, get_free_dof_values(duh₁))
+  get_free_dof_values(duh₂) .= get_free_dof_values(duh₂) .- y_wrk
+
+  # solve for du₂, dh₂
+  ldiv!(Bchol, get_free_dof_values(duh₂))
+
+  # update yⁿ⁺¹
+  get_free_dof_values(y₂) .= get_free_dof_values(y₁) .+ dt .* get_free_dof_values(duh₂)
+end
+
+function new_vtk_step(Ω,file,hn,un,wn)
+  createvtk(Ω,
+            file,
+            cellfields=["hn"=>hn, "un"=>un, "wn"=>wn],
+            nsubcells=4)
+end
+
+function compute_mean_depth!(wrk, L2MM, h)
+  # compute the mean depth over the sphere, for use in the approximate Jacobian
+  mul!(wrk, L2MM, get_free_dof_values(h))
+  h_int = sum(wrk)
+  wrk  .= 1.0
```

Comment:
Replace the next three lines by `a_int=sum(A,dims=[1,2])`

Reply:
done, thanks